Userland/Libraries/LibWeb/HTML/HTMLImageElement.cpp

```diff
@@ -311,55 +314,97 @@ WebIDL::ExceptionOr<JS::NonnullGCPtr<WebIDL::Promise>> HTMLImageElement::decode(
             return true;
         };
 
-        // 2.1 If any of the following are true:
-        // 2.1.1 this's node document is not fully active;
-        // 2.1.1 then reject promise with an "EncodingError" DOMException.
-        if (reject_if_document_not_fully_active())
-            return;
-
-        // 2.1.2  or this's current request's state is broken,
-        // 2.1.2 then reject promise with an "EncodingError" DOMException.
-        if (reject_if_current_request_state_broken())
+        // 2. If any of the following are true:
+        //    - this's node document is not fully active;
+        //    - or this's current request's state is broken,
+        //    then reject promise with an "EncodingError" DOMException.
+        if (reject_if_document_not_fully_active() || reject_if_current_request_state_broken()) {
             return;
+        }
 
-        // 2.2 Otherwise, in parallel wait for one of the following cases to occur, and perform the corresponding actions:
-        Platform::EventLoopPlugin::the().deferred_invoke(JS::create_heap_function(heap(), [this, promise, &realm, reject_if_document_not_fully_active, reject_if_current_request_state_broken] {
+        // 3. Otherwise, in parallel wait for one of the following cases to occur, and perform the corresponding actions:
+        Platform::EventLoopPlugin::the().deferred_invoke(JS::create_heap_function(heap(), [this, promise, &realm, &global] {
+            enum class ImageState {
+                DocumentNoLongerFullyActive,
+                CurrentRequestChanged,
+                CurrentRequestBroken,
+                CurrentRequestCompletelyAvailable,
+            };
+            ImageState image_state;
+            auto previous_state = current_request().state();
             Platform::EventLoopPlugin::the().spin_until(JS::create_heap_function(heap(), [&] {
                 auto state = this->current_request().state();
 
-                return !this->document().is_fully_active() || state == ImageRequest::State::Broken || state == ImageRequest::State::CompletelyAvailable;
-            }));
+                if (state != previous_state) {
+                    image_state = ImageState::CurrentRequestChanged;
```

Comment:
does the state of the current request changing really mean that the current request changed? What about ABA? Should we be storing the "Current request" and making sure the image element still has the same request instead?

Reply:
"or is mutated" :thinking: How do we detect that?